### PR TITLE
Implement quickfix for letting type implement interface

### PR DESCRIPTION
**Pull request by cypressious**
Fixes #KT-11404

### CODE REVIEW

**Comment by mglukhikh** [comment]
Under investigation

**Comment by cypressious** [comment]
@mglukhikh Any feedback? :)

**Comment by cypressious** [comment]
Rebased, removed reference to type, fixed the logic to consider all indirect supertypes, grouped tests and added one

File: idea/src/org/jetbrains/kotlin/idea/quickfix/LetImplementInterfaceFix.kt [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+import org.jetbrains.kotlin.psi.KtPsiFactory
+import org.jetbrains.kotlin.resolve.DescriptorToSourceUtils
+import org.jetbrains.kotlin.resolve.lazy.BodyResolveMode
+import org.jetbrains.kotlin.types.KotlinType
+import org.jetbrains.kotlin.types.typeUtil.isInterface
+import org.jetbrains.kotlin.types.typeUtil.supertypes
+
+private fun KotlinType.renderShort() = IdeDescriptorRenderers.SOURCE_CODE_SHORT_NAMES_IN_TYPES.renderType(this)
+
+private fun KotlinType.getSuperTypesRecursive() = constructor.supertypes.let { it + it.flatMap { it.supertypes() } }
```
Comment (udalov) [inline idea/src/org/jetbrains/kotlin/idea/quickfix/LetImplementInterfaceFix.kt]: This is exponential; please use `TypeUtils#getAllSupertypes` instead

Comment (cypressious) [inline idea/src/org/jetbrains/kotlin/idea/quickfix/LetImplementInterfaceFix.kt]: If I understand correctly, Mikhail is merging this manually in https://github.com/JetBrains/kotlin/commit/b569e267ac165a2dcb0d5c9670b5df4e6c55ae6f#diff-fd98cb60318ca31d01b7fbc6d9f39f3a

Comment (cypressious) [inline idea/src/org/jetbrains/kotlin/idea/quickfix/LetImplementInterfaceFix.kt]: Anyway, fixed it.

**Comment by mglukhikh** [comment]
Pushed to master. I don't think additional code with indirect supertypes is really necessary because we will not have any error and thus no quick-fix. Anyway, I added a test for this case. Sorry for slow consideration.

**Comment by cypressious** [comment]
@mglukhikh The problem arrises when you expect generic type `A<X>` but the type already indirectly implements `A<Y>`. Here's a demonstration:

```kotlin

fun main(args: Array<String>) {
    foo(B()<caret>)
}


fun foo(a: A<Y>) {
}

interface A<T>
open class C : A<X>
class B : C()

class X
class Y
```

**Pull request closed without merging**